### PR TITLE
* init decoupling-post-worker code

### PR DESCRIPTION
**Pull request by gegezai**
Merge based on：
https://github.com/filecoin-project/lotus/discussions/6943

### CODE REVIEW

**Comment by gegezai** [comment]
@jennijuju The code has been debugged and verified, please review the merge

File: types.go [hunk 1]
```diff
@@ -131,3 +132,50 @@ type PrivateSectorInfo struct {
 type AllocationManager interface {
 	Free()
 }
+
+func DeDuplicatePrivateSectorInfo(ctx context.Context, sortPrivSectors *SortedPrivateSectorInfo) (SortedPrivateSectorInfo, error) {
```
Comment (Kubuxu) [inline types.go:136]: Looking at code in Lotus it looks like it should be possible to execute this and split operations before creating the `SortedPrivateSectorInfo`.
I strongly suggest changing it to that.
The SortedPrivateSectorInfo type should only be used for directly passing sector information into FFI not for carrying sector information around or operating on them.

**Review by Kubuxu — CHANGES_REQUESTED** [review]
As requested in the comment, perform operations required first and then create an instance of SortedPrivateSectorInfo.

Changes in proofs.go and distributed.go look good.

**Comment by gegezai** [comment]
@Kubuxu Agree with you. We have a try

**Comment by gegezai** [comment]
@jennijuju  @Kubuxu According to the review comments, the deduplication of SortedPrivateSectorInfo is performed in lotus, because the key member variables of SortedPrivateSectorInfo are private types and cannot be accessed at the lotus layer. 
image

We tried to perform deduplication before calling ffi.NewSortedPrivateSectorInfo and tested it, but proof verification Fail.

Any suggestions for modification?

File: types.go [hunk 2]
```diff
@@ -131,3 +132,50 @@ type PrivateSectorInfo struct {
...
+		flag := false
+		for j := 0; j < new_sector_len-i-1; j++ {
+			if newSortPrivSectors.f[j].SectorNumber > newSortPrivSectors.f[j+1].SectorNumber {
+				newSortPrivSectors.f[j], newSortPrivSectors.f[j+1] = newSortPrivSectors.f[j+1], newSortPrivSectors.f[j]
+				flag = true
+			}
+		}
+		if !flag {
+			break
+		}
```
Comment (Kubuxu) [inline types.go:160]: The above code seems to sort sector infos by SectorNumber, but the original code sorts them by SealedCID, why the change?

Also if you want to sort, please use `sort.Slice`. See types.go#93

Comment (gegezai) [reply types.go:160]: @Kubuxu According to the logic before splitting,Split calculation results need to be sorted by SectorNumber
企业微信截图_16377349048939

Comment (gegezai) [reply types.go:160]: > 

@Kubuxu Any suggestions for modification?

Comment (gegezai) [reply types.go:160]: > 

Because the members of the SortedProvateSectorInfo structure are private and cannot be operated in lotus, we think the DeDuplicatePrivateSectorInfo interface is necessary, or are there any other implementation suggestions? @Kubuxu

Comment (magik6k) [reply types.go:160]: Won't passing sectors deduplicated here to `NewSortedPrivateSectorInfo` sort them correctly? If not, can we change `NewSortedPrivateSectorInfo` to sort correctly?

Comment (gegezai) [reply types.go:160]: @magik6k We can change NewSortedPrivateSectorInfo to sort correctly, If It without affecting the logic of other functions

Comment (gegezai) [reply types.go]: @magik6k Modification has been submitted, please review

File: types.go [hunk 3]
```diff
@@ -89,14 +90,37 @@ func (s *SortedPublicSectorInfo) UnmarshalJSON(b []byte) error {
 
 // NewSortedPrivateSectorInfo returns a SortedPrivateSectorInfo
 func NewSortedPrivateSectorInfo(sectorInfo ...PrivateSectorInfo) SortedPrivateSectorInfo {
-	fn := func(i, j int) bool {
-		return bytes.Compare(sectorInfo[i].SealedCID.Bytes(), sectorInfo[j].SealedCID.Bytes()) == -1
+	var remove_duplicate_privSector = make([]PrivateSectorInfo, 0)
```
Comment (magik6k) [inline types.go:93]: Maybe use a shorter variable name? We're using it in almost every line of this function. (we can also preallocate this slice)

```suggestion
	result := make([]PrivateSectorInfo, 0, len(sectorInfo))
```

File: types.go [hunk 4]
```diff
@@ -89,14 +90,37 @@ func (s *SortedPublicSectorInfo) UnmarshalJSON(b []byte) error {
...
+		}
 	}
 
-	sort.Slice(sectorInfo[:], fn)
+	new_sector_len := len(remove_duplicate_privSector)
+	if new_sector_len < 2 {
+		return SortedPrivateSectorInfo{
+			f: remove_duplicate_privSector,
+		}
+	}
```
[the remark this answers is not in the document]
Comment (gegezai) [reply types.go:107]: Accept

**Comment by gegezai** [comment]
@magik6k  The code has been debugged and verified, please review and merge

**Review by magik6k — APPROVED** [review]
I still don't really understand the change in sorting - why sorting by SealedCID(CommR) worked before, and why we're changing it to sort by SectorNumber. Reading the rust code, the only thing I see is that the code doesn't really seem to care about sorting, at least in case of the monolith post methods.

Will merge this, if it's somehow broken it should be really easy to see that and fix
